docs/design/single-file-unsupported.md

```diff
@@ -0,0 +1,68 @@
+# Single-file unsupported attribute
+
+Starting with .NET Core 3.0 users have the ability to bundle all files used by their applications into a single executable through the publish functionality, `PublishSingleFile`. Although most applications can be agnostic of whether they will be deployed as single-file, there exist some APIs that are known to be incompatible with this deployment mode and thus can cause single-file apps to crash.
+
+While the existing single-file analyzer can help us emit warnings whenever incompatible runtime library APIs are called from within user's code in a project that has been marked with the intent of being published as single-file (through setting `PublishSingleFile` to true in its .csproj), developers can still take a dependency on third-party components implementing functionality that inadvertently make their application no longer compatible with single-file.
+
+## User Experience
+
+For library authors, the new attribute will allow them to easily annotate APIs known to be problematic when used in the context of a single-file application.
+
+```C#
+[SingleFileUnsupported("'Bar' method is not compatible with single-file", "https://help")]
+void Bar()
+{
+    Assembly executingAssembly = Assembly.GetExecutingAssembly();
+    var codeBase = executingAssembly.CodeBase;
+    ...
+}
+```
+
+Consumers of this library who want to publish their application as a single-file binary will commonly do it through setting the corresponding property in the application’s .csproj:
+
+```XML
+<PropertyGroup>
+    <PublishSingleFile>true</PublishSingleFile>
+</PropertyGroup>
+```
+
+If the user now makes use of the problematic API, a warning will be produced:
+
+> File.cs(4,4): Single-file warning ILXXXX: Foo.CallBar(): 'Bar' method is not compatible with single-file. Url.
+```C#
+void CallBar()
+{
+    ...
+    Bar();
+    ...
+}
+```
+The linker will exercise implicit suppression of all other single-file related warnings produced by code within the incompatible method.
+
+If a user is confident that the used API does not pose a problem, the known warning suppression mechanisms can be used.
```

Comment:
Is suppression really the only way to mitigate the warning? What is the syntax to suppress the warning in external NuGet?

Reply:
Not sure how common it will be to try to disable the warning in external NuGet - most developers will simply give up at that point (as it's next to impossible for them to tell if the NuGet has a real issue or not).
This will not be a problem for Roslyn analyzers (since there's no external code being analyzed in that case).
For linker we can use the `ILLink.LinkAttributes.xml` and inject the `UnconditionalSuppressMessageAttribute` into the target NuGet assembly - not easy or simple, but it will work.

Reply:
The are two questions

1. Are suppression and attribute propagation really the only two ways how to resolve this warning?

2. If I'm building a library that has dependency on different library which is not single-file safe how am I going to make my build warning-free?

Reply:
1. Yes - if that experience is not good enough we can definitely try to add some more, but right now that's the plan
2. The warnings are there for a reason - they typically detect a problem with the library. Fixing that does not mean suppression, sometimes it does, but not always. So if the user wants to ignore these, fine, there are well defined warning suppression mechanisms (like NoWarn).

Reply:
Could we include an example how the fix would look like? Is the linker going to recognize code like bellow as safe or do I have to suppress it?

```c#
static string GetPath()
{
	try {
		return typeof(object).Assembly.CodeBase;
	} catch {
		return "";
	}
}
```

Reply:
I've added an example that illustrates the scenario you point out. The linker (or global analyzer) will not recognize that code as safe, you will have to manually suppress it. I think this is a little bit of a bummer and would definitely like to have an API to guard calls to methods that are not compatible with a specific capability (e.g., single-file), but I think that's out of scope for this doc.

Reply:
I think without capabilities API developers will simply suppress the warning as there is no easy way to add special handling and we'll be where we are today.